lib/Transforms/Scalar/DxilLoopUnroll.cpp

```diff
@@ -467,6 +487,9 @@ static void FindProblemBlocks(BasicBlock *Header, const SmallVectorImpl<BasicBlo
       //
       if (hlsl::dxilutil::IsHLSLObjectType(EltType)) {
         ProblemBlocks.insert(GEP->getParent());
+        if (AllocaInst *AI = GetAllocaFromGEP(cast<GEPOperator>(GEP))) {
```

Comment:
I think we should handle the one case where you're indexing a global resource array with external linkage (which is perfectly legal).
GetAllocaFromGEP could be changed to GetGEPPtrOrigin or something like that.  It would return either an AllocaInst or a GlobalVariable.  If GlobalVariable, and GV.getLinkage() == GlobalVariable::LinkageTypes::ExternalLinkage, skip it.  Otherwise you mark a problem block, and if it's an AllocaInst, add that to the ProblemAllocas.